File: backend/api/routes/repository_tree.py
"""
Repository Tree API Routes
Handles repository folder structure operations for workspace analysis
Completely separate from workflow treeview API
"""

from fastapi import APIRouter, Depends, HTTPException, status
from typing import Dict, List, Any, Optional
from pydantic import BaseModel
import logging
from sqlalchemy import select

from database.config import db_manager
from core.repository_tree_manager import RepositoryTreeManager
from core.security import get_current_user
from database.models import Organization, User

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/repository-tree", tags=["Repository Tree"])


# =============================================================================
# Helper Functions
# =============================================================================

async def get_user_uuid_from_auth_id(session, auth_user_id: str) -> str:
    """
    Get user UUID from Auth0 user ID
    
    Args:
        session: Database session
        auth_user_id: Auth0 user ID (e.g., 'google-oauth2|123')
    
    Returns:
        User UUID string
        
    Raises:
        HTTPException: If user not found
    """
    result = await session.execute(
        select(User).filter(User.auth_user_id == auth_user_id)
    )
    user = result.scalar_one_or_none()
    
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"User not found in database. Please log in again."
        )
    
    return user.id


# =============================================================================
# Pydantic Models for Request/Response
# =============================================================================

class RepositoryTreeSaveRequest(BaseModel):
    """Request model for saving repository tree"""
    organization_login: str
    tree_data: List[Dict[str, Any]]
    name: Optional[str] = "Repository Structure"
    description: Optional[str] = None


class RepositoryTreeResponse(BaseModel):
    """Response model for repository tree operations"""
    success: bool
    data: Optional[List[Dict[str, Any]]] = None
    metadata: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    message: Optional[str] = None


# =============================================================================
# API Endpoints
# =============================================================================

@router.get("/{org_login}")
async def get_repository_tree(
    org_login: str,
    current_user: str = Depends(get_current_user)
):
    """
    Get repository tree structure for an organization
    
    GET /api/repository-tree/{org_login}
    
    Returns:
        Repository tree data with folders and repositories
    """
    try:
        auth_user_id = current_user  # Auth0 user ID
        
        if not auth_user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not authenticated"
            )
        
        # Get organization from database
        async with db_manager.get_session() as session:
            # Get user UUID from Auth0 ID
            user_id = await get_user_uuid_from_auth_id(session, auth_user_id)
            
            result = await session.execute(
                select(Organization).filter(Organization.login == org_login)
            )
            org = result.scalar_one_or_none()
            
            if not org:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Organization '{org_login}' not found"
                )
            
            # Get repository tree
            manager = RepositoryTreeManager(session)
            tree_result = await manager.get_repository_tree(org.id, user_id)
            
            if tree_result.get("success"):
                return {
                    "success": True,
                    "data": tree_result.get("data", []),
                    "metadata": tree_result.get("metadata", {})
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail=tree_result.get("error", "Failed to get repository tree")
                )
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in get_repository_tree: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.post("/save")
async def save_repository_tree(
    request: RepositoryTreeSaveRequest,
    current_user: str = Depends(get_current_user)
):
    """
    Save or update repository tree structure
    
    POST /api/repository-tree/save
    Body: {
        "organization_login": "org-name",
        "tree_data": [...],
        "name": "Repository Structure",
        "description": "Optional description"
    }
    
    Returns:
        Success status and tree ID
    """
    try:
        logger.info(f"📥 Save repository tree request: org={request.organization_login}, data_length={len(request.tree_data)}")
        
        auth_user_id = current_user  # Auth0 user ID
        
        if not auth_user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not authenticated"
            )
        
        # Get organization from database
        async with db_manager.get_session() as session:
            # Get user UUID from Auth0 ID
            user_id = await get_user_uuid_from_auth_id(session, auth_user_id)
            
            result = await session.execute(
                select(Organization).filter(Organization.login == request.organization_login)
            )
            org = result.scalar_one_or_none()
            
            if not org:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Organization '{request.organization_login}' not found"
                )
            
            # Save repository tree
            manager = RepositoryTreeManager(session)
            save_result = await manager.save_repository_tree(
                organization_id=org.id,
                user_id=user_id,
                tree_data=request.tree_data,
                name=request.name,
                description=request.description
            )
            
            if save_result.get("success"):
                return {
                    "success": True,
                    "tree_id": save_result.get("tree_id"),
                    "version": save_result.get("version"),
                    "message": save_result.get("message")
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail=save_result.get("error", "Failed to save repository tree")
                )
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in save_repository_tree: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.delete("/{org_login}")
async def delete_repository_tree(
    org_login: str,
    current_user: str = Depends(get_current_user)
):
    """
    Delete repository tree structure
    
    DELETE /api/repository-tree/{org_login}
    
    Returns:
        Success status
    """
    try:
        auth_user_id = current_user  # Auth0 user ID
        
        if not auth_user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not authenticated"
            )
        
        # Get organization from database
        async with db_manager.get_session() as session:
            # Get user UUID from Auth0 ID
            user_id = await get_user_uuid_from_auth_id(session, auth_user_id)
            
            result = await session.execute(
                select(Organization).filter(Organization.login == org_login)
            )
            org = result.scalar_one_or_none()
            
            if not org:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Organization '{org_login}' not found"
                )
            
            # Delete repository tree
            manager = RepositoryTreeManager(session)
            delete_result = await manager.delete_repository_tree(org.id, user_id)
            
            if delete_result.get("success"):
                return {
                    "success": True,
                    "message": delete_result.get("message")
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail=delete_result.get("error", "Failed to delete repository tree")
                )
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in delete_repository_tree: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )


@router.get("/{org_login}/statistics")
async def get_repository_tree_statistics(
    org_login: str,
    current_user: str = Depends(get_current_user)
):
    """
    Get statistics about repository tree structure
    
    GET /api/repository-tree/{org_login}/statistics
    
    Returns:
        Statistics: folder count, repository count, workflow count
    """
    try:
        auth_user_id = current_user  # Auth0 user ID
        
        if not auth_user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not authenticated"
            )
        
        # Get organization from database
        async with db_manager.get_session() as session:
            # Get user UUID from Auth0 ID
            user_id = await get_user_uuid_from_auth_id(session, auth_user_id)
            
            result = await session.execute(
                select(Organization).filter(Organization.login == org_login)
            )
            org = result.scalar_one_or_none()
            
            if not org:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Organization '{org_login}' not found"
                )
            
            # Get statistics
            manager = RepositoryTreeManager(session)
            stats_result = await manager.get_tree_statistics(org.id, user_id)
            
            if stats_result.get("success"):
                return {
                    "success": True,
                    "statistics": stats_result.get("statistics", {})
                }
            else:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail=stats_result.get("error", "Failed to get statistics")
                )
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in get_repository_tree_statistics: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e)
        )

File: frontend/src/lib/repositoryTree.js
/**
 * Repository Tree Client
 * Handles repository folder structure for workspace analysis and DevSecOps intelligence
 *
 * COMPLETELY SEPARATE from workflow treeview (github.js ProjectTree methods)
 *
 * Purpose:
 * - Organize repositories in folder structures
 * - Future: Workspace analysis and DevSecOps maturity assessment
 * - Future: AI-powered queries and insights
 * - Future: OWASP DSOMM compliance tracking
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:9000';

class RepositoryTreeClient {
	constructor() {
		this.baseUrl = API_BASE_URL;
		console.log('🌲 Repository Tree Client initialized (separate from workflow treeview)');
	}

	/**
	 * Get authentication token from localStorage or sessionStorage
	 */
	getAuthToken() {
		if (typeof window === 'undefined') return null;

		return (
			localStorage.getItem('auth_token') ||
			sessionStorage.getItem('auth_token') ||
			localStorage.getItem('github_token')
		);
	}

	/**
	 * Get repository tree structure for an organization
	 *
	 * @param {string} orgName - Organization name/login
	 * @returns {Promise<{success: boolean, data: Array, metadata: Object, error?: string}>}
	 */
	async getRepositoryTree(orgName) {
		try {
			console.log(`📦 Fetching repository tree for organization: ${orgName}`);

			const response = await fetch(`${this.baseUrl}/api/repository-tree/${orgName}`, {
				method: 'GET',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${this.getAuthToken()}`
				}
			});

			if (!response.ok) {
				const errorData = await response.json().catch(() => ({}));
				throw new Error(errorData.detail || errorData.error || 'Failed to get repository tree');
			}

			const data = await response.json();
			console.log(`✅ Repository tree loaded:`, data);

			return {
				success: true,
				data: data.data || [],
				metadata: data.metadata || {}
			};
		} catch (error) {
			console.error('❌ Error getting repository tree:', error);
			return {
				success: false,
				data: [],
				metadata: {},
				error: error.message || 'Failed to get repository tree'
			};
		}
	}

	/**
	 * Save or update repository tree structure
	 *
	 * @param {string} orgName - Organization name/login
	 * @param {Array} treeData - Tree structure with folders and repositories
	 * @param {string} [name] - Optional tree name
	 * @param {string} [description] - Optional description
	 * @returns {Promise<{success: boolean, tree_id?: string, version?: number, message?: string, error?: string}>}
	 */
	async saveRepositoryTree(orgName, treeData, name = 'Repository Structure', description = null) {
		try {
			console.log(`💾 Saving repository tree for organization: ${orgName}`, {
				folders: this._countNodesByType(treeData, 'folder'),
				repositories: this._countNodesByType(treeData, 'repository'),
				workflows: this._countNodesByType(treeData, 'workflow')
			});

			const response = await fetch(`${this.baseUrl}/api/repository-tree/save`, {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${this.getAuthToken()}`
				},
				body: JSON.stringify({
					organization_login: orgName,
					tree_data: treeData,
					name: name,
					description: description
				})
			});

			if (!response.ok) {
				const errorData = await response.json().catch(() => ({}));
				throw new Error(errorData.detail || errorData.error || 'Failed to save repository tree');
			}

			const data = await response.json();
			console.log(`✅ Repository tree saved successfully:`, data);

			return {
				success: true,
				tree_id: data.tree_id,
				version: data.version,
				message: data.message || 'Repository tree saved successfully'
			};
		} catch (error) {
			console.error('❌ Error saving repository tree:', error);
			return {
				success: false,
				error: error.message || 'Failed to save repository tree'
			};
		}
	}

	/**
	 * Delete repository tree structure
	 *
	 * @param {string} orgName - Organization name/login
	 * @returns {Promise<{success: boolean, message?: string, error?: string}>}
	 */
	async deleteRepositoryTree(orgName) {
		try {
			console.log(`🗑️ Deleting repository tree for organization: ${orgName}`);

			const response = await fetch(`${this.baseUrl}/api/repository-tree/${orgName}`, {
				method: 'DELETE',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${this.getAuthToken()}`
				}
			});

			if (!response.ok) {
				const errorData = await response.json().catch(() => ({}));
				throw new Error(errorData.detail || errorData.error || 'Failed to delete repository tree');
			}

			const data = await response.json();
			console.log(`✅ Repository tree deleted successfully`);

			return {
				success: true,
				message: data.message || 'Repository tree deleted successfully'
			};
		} catch (error) {
			console.error('❌ Error deleting repository tree:', error);
			return {
				success: false,
				error: error.message || 'Failed to delete repository tree'
			};
		}
	}

	/**
	 * Get repository tree statistics
	 *
	 * @param {string} orgName - Organization name/login
	 * @returns {Promise<{success: boolean, statistics?: Object, error?: string}>}
	 */
	async getStatistics(orgName) {
		try {
			console.log(`📊 Fetching repository tree statistics for: ${orgName}`);

			const response = await fetch(`${this.baseUrl}/api/repository-tree/${orgName}/statistics`, {
				method: 'GET',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${this.getAuthToken()}`
				}
			});

			if (!response.ok) {
				const errorData = await response.json().catch(() => ({}));
				throw new Error(errorData.detail || errorData.error || 'Failed to get statistics');
			}

			const data = await response.json();
			console.log(`✅ Statistics loaded:`, data.statistics);

			return {
				success: true,
				statistics: data.statistics || {
					total_folders: 0,
					total_repositories: 0,
					total_workflows: 0,
					private_repos: 0,
					public_repos: 0
				}
			};
		} catch (error) {
			console.error('❌ Error getting statistics:', error);
			return {
				success: false,
				statistics: {
					total_folders: 0,
					total_repositories: 0,
					total_workflows: 0,
					private_repos: 0,
					public_repos: 0
				},
				error: error.message || 'Failed to get statistics'
			};
		}
	}

	/**
	 * Helper: Count nodes by type in tree structure
	 * @private
	 */
	_countNodesByType(treeData, type) {
		let count = 0;

		function traverse(nodes) {
			for (const node of nodes) {
				if (node.type === type) {
					count++;
				}
				if (node.children) {
					traverse(node.children);
				}
			}
		}

		traverse(treeData);
		return count;
	}

	/**
	 * Validate tree structure
	 * Ensures all nodes have required fields
	 *
	 * @param {Array} treeData - Tree structure to validate
	 * @returns {boolean} True if valid
	 */
	validateTreeStructure(treeData) {
		if (!Array.isArray(treeData)) {
			console.error('❌ Tree data must be an array');
			return false;
		}

		function validateNode(node, path = 'root') {
			// Required fields
			if (!node.id || !node.name || !node.type) {
				console.error(`❌ Invalid node at ${path}: missing id, name, or type`, node);
				return false;
			}

			// Valid types
			const validTypes = ['folder', 'repository', 'workflow'];
			if (!validTypes.includes(node.type)) {
				console.error(`❌ Invalid node type at ${path}: ${node.type}`, node);
				return false;
			}

			// Validate children recursively
			if (node.children) {
				if (!Array.isArray(node.children)) {
					console.error(`❌ Invalid children at ${path}: must be an array`, node);
					return false;
				}

				for (let i = 0; i < node.children.length; i++) {
					if (!validateNode(node.children[i], `${path}.${node.name}[${i}]`)) {
						return false;
					}
				}
			}

			return true;
		}

		for (let i = 0; i < treeData.length; i++) {
			if (!validateNode(treeData[i], `root[${i}]`)) {
				return false;
			}
		}

		console.log('✅ Tree structure is valid');
		return true;
	}

	/**
	 * Get flattened list of all folders in tree
	 * Useful for folder selection dropdowns
	 *
	 * @param {Array} treeData - Tree structure
	 * @param {number} [depth=0] - Current depth (for indentation)
	 * @returns {Array} Flattened list of folders
	 */
	getFlattenedFolders(treeData, depth = 0) {
		let result = [];

		for (const node of treeData) {
			if (node.type === 'folder') {
				result.push({ ...node, depth });

				if (node.children) {
					result.push(...this.getFlattenedFolders(node.children, depth + 1));
				}
			}
		}

		return result;
	}

	/**
	 * Find a node by ID in tree structure
	 *
	 * @param {Array} treeData - Tree structure
	 * @param {string} nodeId - Node ID to find
	 * @returns {Object|null} Found node or null
	 */
	findNodeById(treeData, nodeId) {
		function search(nodes) {
			for (const node of nodes) {
				if (node.id === nodeId) {
					return node;
				}
				if (node.children) {
					const found = search(node.children);
					if (found) return found;
				}
			}
			return null;
		}

		return search(treeData);
	}

	/**
	 * Count total items in a node (including nested items)
	 *
	 * @param {Object} node - Node to count
	 * @returns {number} Total item count
	 */
	countItemsInNode(node) {
		if (!node || !node.children) return 0;

		let count = 0;
		for (const child of node.children) {
			count++;
			if (child.children) {
				count += this.countItemsInNode(child);
			}
		}
		return count;
	}
}

// Export singleton instance
export const repositoryTreeClient = new RepositoryTreeClient();

// Make available globally for testing and debugging
if (typeof window !== 'undefined') {
	window.repositoryTreeClient = repositoryTreeClient;
	console.log(
		'🧪 Repository Tree client available globally as window.repositoryTreeClient for testing'
	);
}

export default repositoryTreeClient;
